File: python/eiko/animate_eikonal.py
import sys

def animate_eikonal(u, v=1.0, color_map='gray', video_filename='', 
                    title='EIKONAL WAVEFRONT', pulse_width=80.0, 
                    speed=0.5, overlay=None, outline=None, 
                    style='real', render_mode='slice',
                    extent=None):
    """
    ANIMATE_EIKONAL Visualizes Eikonal equation travel times as an animated wave.
    """
    
    # Required dependencies to run the visualization.
    try:
        import numpy as np
        import matplotlib.pyplot as plt
        from matplotlib.animation import FuncAnimation
    except ImportError:
        raise ImportError(
            "The animation utility requires 'numpy' and 'matplotlib'. "
            "Please install it using: pip install \"eiko[examples]\""
        )
    
    # Optional dependency for contours/isosurfaces
    try:
        from skimage import measure
        HAS_SKIMAGE = True
    except ImportError:
        HAS_SKIMAGE = False
        print("[Eiko] Warning: 'scikit-image' not found. Contours, overlays, and 3D isosurfaces will be disabled.", file=sys.stderr)
    
    # --- Data Type & Device Checks (e.g., PyTorch -> Numpy) ---
    if hasattr(u, 'cpu'): u = u.detach().cpu().numpy()
    if hasattr(v, 'cpu'): v = v.detach().cpu().numpy()
    if hasattr(overlay, 'cpu'): overlay = overlay.detach().cpu().numpy()
    if hasattr(outline, 'cpu'): outline = outline.detach().cpu().numpy()

    is_3d = u.ndim == 3
    
    # --- Safe Data Initialization ---
    valid_u = u[np.isfinite(u)]
    if len(valid_u) == 0:
        print("Warning: Travel time field u is completely Inf. Nothing to animate.", file=sys.stderr)
        return

    maxv = np.max(valid_u)
    u_safe = np.copy(u).astype(np.float32)
    u_safe[np.isinf(u_safe)] = maxv + pulse_width

    # Handle Permutation and setup for 3D
    if is_3d:
        # In Numpy, typical 3D ordering is [Z, Y, X]. We keep it as is, 
        # but map slices to dimensions.
        Nz, Ny, Nx = u.shape
        v_anim = v
    else:
        v_anim = v
        Ny, Nx = u.shape

    # --- Figure & Axis Setup ---
    fig = plt.figure(figsize=(12, 8), facecolor='white')
    
    if style.lower() == 'real':
        clim = [-1, 1]
    elif style.lower() == 'abs':
        clim = [0, 1]
    elif style.lower() == 'db':
        clim = [-30, 0]
        db_ref = 20 * np.log10(1.0 + 1e-12) # +1e-12 prevents log10(0)

    has_extent = extent is not None
    unit_suffix = ' [mm]' if has_extent else ''

    # Setup Render Objects
    if is_3d:
        # 3D Axes
        ax = fig.add_subplot(111, projection='3d')
        ax.set_title(title, fontsize=16, fontweight='bold')
        
        ax.set_xlabel(f'Lateral - X{unit_suffix}', fontsize=11, labelpad=10)
        ax.set_ylabel(f'Elevation - Y{unit_suffix}', fontsize=11, labelpad=10)
        ax.set_zlabel(f'Depth - Z{unit_suffix}', fontsize=11, labelpad=10)
        ax.invert_zaxis() # Ultrasound convention: Depth increases downwards
        
        # Fix inside-out tick directions by forcing ticks outward
        for axis in (ax.xaxis, ax.yaxis, ax.zaxis):
            axis._axinfo['tick']['inward_factor'] = 0.0
            axis._axinfo['tick']['outward_factor'] = 0.3
        
        if len(extent) == 6:
            xmin, xmax, ymin, ymax, zmin, zmax = extent
            ax.set_xlim(xmin, xmax)
            ax.set_ylim(ymin, ymax)
            ax.set_zlim(zmax, zmin) # Inverted for depth
            
            # Enforce true 1:1:1 physical aspect ratio based on domain size
            ax.set_box_aspect((xmax - xmin, ymax - ymin, abs(zmax - zmin)))
            
            lat_vals = np.linspace(xmin, xmax, Nx)
            elev_vals = np.linspace(ymin, ymax, Ny)
            depth_vals = np.linspace(zmin, zmax, Nz)
        else:
            lat_vals = np.arange(Nx)
            elev_vals = np.arange(Ny)
            depth_vals = np.arange(Nz)
        
        sz, sy, sx = Nz//2, Ny//2, Nx//2
        iso_val = -6 if style.lower() == 'db' else 0.5
        
    else:
        # 2D Rendering
        ax = fig.add_subplot(111)
        ax.set_title(title, fontsize=16, fontweight='bold', fontname='serif')
        
        ax.set_xlabel(f'Lateral - X{unit_suffix}', fontsize=14, fontname='serif')
        ax.set_ylabel(f'Depth - Z{unit_suffix}', fontsize=14, fontname='serif')
        
        ax.tick_params(direction='out')
        
        if len(extent) >= 4:
            xmin, xmax, ymin, ymax = extent[:4]
            im = ax.imshow(np.zeros((Ny, Nx)), vmin=clim[0], vmax=clim[1], cmap=color_map, aspect='equal', extent=[xmin, xmax, ymax, ymin], origin='lower')
        else:
            im = ax.imshow(np.zeros((Ny, Nx)), vmin=clim[0], vmax=clim[1], cmap=color_map, aspect='equal')
        
        # In matplotlib, imshow puts (0,0) at top-left, which inherently matches 
        # the ultrasound convention (Depth going down).
        
        if outline is not None and HAS_SKIMAGE:
            ax.contour(outline, levels=[0.5], colors='k', linewidths=2)
            
        if overlay is not None and HAS_SKIMAGE:
            contours = measure.find_contours(overlay, 0.5)
            for contour in contours:
                # contour is (row, col) -> (Y, X)
                ax.fill(contour[:, 1], contour[:, 0], color='red', alpha=0.3, edgecolor='r', linewidth=2.5)

    freq = 6 * np.pi
    
    # --- Animation Update Logic ---
    def compute_frame(t):
        """Calculates the analytical wave packet for a given time step."""
        diff_t = u_safe - t
        valid_mask = np.abs(diff_t) <= (pulse_width / 2.0)
        
        img = np.zeros_like(u_safe, dtype=np.complex64)
        
        if np.any(valid_mask):
            d_valid = diff_t[valid_mask]
            
            # Analytical Hanning envelope
            envelope = 0.5 * (1.0 + np.cos(2 * np.pi * d_valid / pulse_width))
            
            # Apply envelope and carrier phase
            img[valid_mask] = envelope * np.exp(1j * freq * d_valid / pulse_width)
            
        img = img * v_anim
        
        # Apply Style
        if style.lower() == 'real':
            final_img = np.real(img)
        elif style.lower() == 'abs':
            final_img = np.abs(img)
        elif style.lower() == 'db':
            final_img = 20 * np.log10(np.abs(img) + 1e-12) - db_ref
            
        # Make zero/inactive amplitude regions transparent using a masked array
        # final_img = np.ma.masked_where(np.abs(img) < 1e-5, final_img)
            
        return final_img

    # time_steps = np.arange(0, maxv + pulse_width, speed)
    start_time = np.min(u_safe)
    time_steps = np.arange(start_time, maxv + pulse_width, speed)
    
    def update(frame_idx):
        t = time_steps[frame_idx]
        final_img = compute_frame(t)
        
        #if frame_idx % 20 == 0: # Check every 20 frames to avoid spamming
        #    print(f"Frame {frame_idx} (t={t:.1f}): Data min/max = {final_img.min():.3f} / {final_img.max():.3f}")

        if is_3d:
            # Clear all collections (slices or isosurfaces) from the previous frame.
            for coll in ax.collections:
                coll.remove()
            
            # Define levels to force contouring (20 steps is usually enough for smooth waves)
            levels = np.linspace(clim[0], clim[1], 20)
            if render_mode == 'slice':
                X_lat, Y_elev = np.meshgrid(lat_vals, elev_vals)
                ax.contourf(X_lat, Y_elev, final_img[sz, :, :], zdir='z', offset=depth_vals[sz], cmap=color_map, vmin=clim[0], vmax=clim[1], levels=levels)
                
                Y_elev, Z_dep = np.meshgrid(elev_vals, depth_vals)
                ax.contourf(final_img[:, :, sx], Y_elev, Z_dep, zdir='x', offset=lat_vals[sx], cmap=color_map, vmin=clim[0], vmax=clim[1], levels=levels)
                
                X_lat, Z_dep = np.meshgrid(lat_vals, depth_vals)
                ax.contourf(X_lat, final_img[:, sy, :], Z_dep, zdir='y', offset=elev_vals[sy], cmap=color_map, vmin=clim[0], vmax=clim[1], levels=levels)
                
            elif render_mode == 'isosurface' and HAS_SKIMAGE:
                try:
                    # Use marching cubes to find the isosurface.
                    verts, faces, _, _ = measure.marching_cubes(final_img, level=iso_val)
                    from mpl_toolkits.mplot3d.art3d import Poly3DCollection
                    mesh = Poly3DCollection(verts[faces], alpha=0.7)
                    mesh.set_facecolor([0.2, 0.6, 1.0])
                    mesh.set_edgecolor('none')
                    ax.add_collection3d(mesh)
                except ValueError:
                    pass # Fails cleanly if the volume doesn't contain the isovalue yet
            
            return ax.collections
        else:
            im.set_data(final_img)
            return [im]

    # --- Run / Export Animation ---
    # Blit=True massively speeds up 2D rendering by only redrawing changed pixels.
    # Blit is not well-supported for matplotlib 3D axes.
    anim = FuncAnimation(fig, update, frames=len(time_steps), 
                                   interval=1000/60, blit=not is_3d)

    if video_filename:
        print(f"Exporting video to {video_filename} (this may take a moment)...")
        # Ensure you have ffmpeg installed on your Ubuntu machine (sudo apt install ffmpeg)
        writer = animation.FFMpegWriter(fps=60, bitrate=2000)
        anim.save(video_filename, writer=writer)
        print("Video export complete.")
    else:
        plt.show()

    return anim

File: matlab/eiko/animate_eikonal.m
function plt = animate_eikonal(u, varargin)
% ANIMATE_EIKONAL Visualizes Eikonal equation travel times as an animated wave.
%
% Usage:
%   animate_eikonal(u)
%   animate_eikonal(u, v, 'Style', 'db', 'VideoFilename', 'out.mp4')
%
% Required Inputs:
%   u - 2D or 3D matrix of travel times (can contain Inf for unreachable areas)
%
% Optional Inputs:
%   v - ND matrix representing amplitude/direction field v. Default is 1.
%
% Optional Name-Value Pairs:
%   'Extent'        - [xmin, xmax, zmin, zmax] (2D) or [xmin, xmax, ymin, ymax, zmin, zmax] (3D). Maps indices to physical units.
%   'VelocityField' - 2D matrix (H x W) representing the direction field v. 
%                     If provided, a faint streamline/quiver overlay is added.
%   'VideoFilename' - String. If provided (e.g., 'wave.mp4'), exports the animation.
%   'Title'         - String to display above the animation.
%   'PulseWidth'    - Spatial width of the wave packet (default: 80).
%   'Speed'         - Time step per frame (default: 0.5). Lower is smoother.
%   'Outline'       - Outline to overlay on top of the 2D/3D image.
%   'Overlay'       - Overlay to put on top of the 2D image.
%   'Style'         - 'real', 'abs', or 'db' (default: 'real').
%   'RenderMode'    - For 3D only: 'slice', 'isosurface', 'contourslice', 'cloud', or 'volshow' (default: 'slice').

    % --- Input Parsing ---
    p = inputParser;
    addRequired(p, 'u', @isnumeric);
    addOptional(p, 'v', 1, @isnumeric);
    addParameter(p, 'ColorMap', gray(256), @isnumeric);
    addParameter(p, 'VideoFilename', '', @ischar);
    addParameter(p, 'Title', 'EIKONAL WAVEFRONT', @ischar);
    addParameter(p, 'PulseWidth', 80, @isnumeric);
    addParameter(p, 'Speed', 0.5, @isnumeric);
    addParameter(p, 'Overlay', [], @(x) isnumeric(x) || islogical(x));
    addParameter(p, 'Outline', [], @(x) isnumeric(x) || islogical(x));
    addParameter(p, 'Style', 'real', @ischar);  % MUSTBEMEMBER
    addParameter(p, 'RenderMode', 'slice', @ischar);
    addParameter(p, 'Extent', [], @isnumeric);

    parse(p, u, varargin{:});
    opts = p.Results;

    % Check dimensionality
    is3D = ndims(u) == 3;

    % Format units if an Extent was passed
    has_extent = ~isempty(opts.Extent);
    if has_extent
        unit_suffix = ' [mm]';
    else
        unit_suffix = '';
    end

    % --- Safe Data Initialization ---
    maxv = gather(max(u(u < inf))); 
    if isempty(maxv) || isinf(maxv)
        warning('Travel time field u is completely Inf. Nothing to animate.');
        return;
    end
    
    if isempty(opts.v)
        opts.v = 1;
    end

    u_safe = single(u); 
    u_safe(isinf(u_safe)) = maxv + opts.PulseWidth; % Push boundaries far away

    % Handle Permutation for 3D (to fix the XY plane transposition bug)
    if is3D
        % Permute from [Z, X, Y] -> [Y, X, Z] for MATLAB's meshgrid standard
        u_safe = permute(u_safe, [3, 2, 1]);
        if ~isscalar(opts.v)
            v_anim = permute(opts.v, [3, 2, 1]);
        else
            v_anim = opts.v;
        end
        Nx = size(u, 2); % Lateral dimension
        Ny = size(u, 3); % Elevation dimension
        Nz = size(u, 1); % Depth dimension

        if has_extent && numel(opts.Extent) == 6
            % Map Extent based on original array: [Depth, Lateral, Elevation]
            x_vals = linspace(opts.Extent(3), opts.Extent(4), Nx); % Lateral
            y_vals = linspace(opts.Extent(5), opts.Extent(6), Ny); % Elevation
            z_vals = linspace(opts.Extent(1), opts.Extent(2), Nz); % Depth
        else
            x_vals = 1:Nx;
            y_vals = 1:Ny;
            z_vals = 1:Nz;
        end

        [X_grid, Y_grid, Z_grid] = meshgrid(x_vals, y_vals, z_vals);
    else
        v_anim = opts.v;
    end

    % --- Setup Gabor Wave Packet Frequency ---
    freq = 6 * pi;     

    % --- Figure & Axis Setup ---
    fig = figure('Color', [1 1 1], 'Position', [100, 100, size(u,2)*1.5*1.5, size(u,1)*1.5*1.5]); 
    set(fig, 'InvertHardcopy', 'off'); 
    
    ax = gca;

    if strcmpi(opts.Style, 'real')
        clim = [-1, 1];
    elseif strcmpi(opts.Style, 'abs')
        clim = [0, 1];
    elseif strcmpi(opts.Style, 'db')
        clim = [-30, 0];
        db_ref = 20 * log10(abs(1));
    end

    % Setup Render Objects
    if is3D
        % Calculate mid-plane slices
        sx_idx = round(Nx/2);
        sy_idx = round(Ny/2);
        sz_idx = round(Nz/2);
        
        sx = x_vals(sx_idx);
        sy = y_vals(sy_idx);
        sz = z_vals(sz_idx);

        % Calculate a threshold for the expanding bubble
        if strcmpi(opts.Style, 'db')
            iso_val = -6; % -6 dB envelope
        else
            iso_val = 0.5; % 50% amplitude
        end

        if strcmpi(opts.RenderMode, 'slice')
            hSlices = slice(ax, X_grid, Y_grid, Z_grid, zeros(size(u_safe)), sx, sy, sz, 'cubic');
            % FaceAlpha set to 'interp' allowing dynamic AlphaData transparency later
            set(hSlices, 'EdgeColor', 'none', 'FaceAlpha', 'interp');
            shading(ax, 'interp');
        elseif strcmpi(opts.RenderMode, 'isosurface')
            % Create an empty patch object
            hPatch = patch(ax, 'Faces', [], 'Vertices', [], ...
            'FaceColor', [0.2, 0.6, 1.0], 'EdgeColor', 'none', ...
            'FaceAlpha', 0.7, 'FaceLighting', 'gouraud', 'BackFaceLighting', 'reverselit');
            camlight('right'); lighting gouraud;
        elseif strcmpi(opts.RenderMode, 'contourslice')
            % Initialize empty handle array for contours
            hContours = [];
        elseif strcmpi(opts.RenderMode, 'cloud')
            cloud_thresh = iso_val * 0.4; % Lower threshold to grab more wave volume
            % Initialize empty scatter3 plot
            hCloud = scatter3(ax, NaN, NaN, NaN, 15, NaN, 'filled', ...
                'MarkerFaceAlpha', 0.15, 'MarkerEdgeColor', 'none');
        elseif strcmpi(opts.RenderMode, 'volshow')
            if exist('volshow', 'file') ~= 2
                error('The ''volshow'' RenderMode requires the Image Processing Toolbox.');
            end
            % Initialize volume viewer object natively inside our axes
            ax = [];
            viewer = viewer3d(fig, 'BackgroundColor', [1 1 1]);
            viewer.CameraUpVector = [0, 0, -1];
            hVol = volshow(zeros(size(u_safe), 'single'), 'Parent', viewer);
            hVol.Colormap = opts.ColorMap;
            % Cubic alphamap: Low values are completely invisible, high values are solid
            hVol.Alphamap = linspace(0, 1, 256).^3; 
        end

        if not(isempty(ax))
            % Labels to verify physical axes
            xlabel(ax, ['X (Lateral)' unit_suffix], 'Interpreter', 'latex', 'FontSize', 18);
            ylabel(ax, ['Y (Elevation)' unit_suffix], 'Interpreter', 'latex', 'FontSize', 18);
            zlabel(ax, ['Z (Depth)' unit_suffix], 'Interpreter', 'latex', 'FontSize', 18);
    
            % Ultrasound convention: Depth increases downwards
            set(ax, 'ZDir', 'reverse', 'TickLabelInterpreter', 'latex', 'FontSize', 14, ...
                'FontName', 'Times New Roman', 'TickDir', 'out', 'Color', [0.5 0.5 0.5]);
    
            view(3);
            if has_extent
                xlim(ax, [min(x_vals), max(x_vals)]); 
                ylim(ax, [min(y_vals), max(y_vals)]); 
                zlim(ax, [min(z_vals), max(z_vals)]);
            else
                xlim(ax, [1, Nx]); ylim(ax, [1, Ny]); zlim(ax, [1, Nz]);
            end
            
            daspect(ax, [1 1 1]); % Strictly respects manual physical limits for true 1:1:1 aspect
            axis vis3d  % Freeze aspect ratio during rotation
            grid on;
    
            ax.XAxis.Color = 'k';
            ax.YAxis.Color = 'k';
            ax.ZAxis.Color = 'k';
    
            if ~strcmpi(opts.RenderMode, 'volshow')
                colormap(ax, opts.ColorMap);
                caxis(ax, clim); % Use clim(ax, clim) if on MATLAB R2022a or newer.

                % --- 3D Outline Handling ---
                if ~isempty(opts.Outline)
                    % Ensure Outline is also permuted to match our [Y, X, Z] volume
                    outline_3d = permute(logical(opts.Outline), [3, 2, 1]);
                    
                    % Generate isosurface for the outline
                    fv = isosurface(X_grid, Y_grid, Z_grid, double(outline_3d), 0.5);
                    
                    % Add as a patch object, restoring the clean red/pink original aesthetic
                    if ~isempty(fv.faces)
                        patch(ax, 'Faces', fv.faces, 'Vertices', fv.vertices, ...
                              'FaceColor', [1.0 0.8 0.8], ...
                              'EdgeColor', 'none', ...
                              'FaceAlpha', 0.3, ...
                              'FaceLighting', 'gouraud', ...
                              'BackFaceLighting', 'reverselit');
                    end
                end
            end
        end
    else
        % 2D Rendering: imagesc naturally places Dim 1 on the Y-axis going down.
        if has_extent && numel(opts.Extent) >= 4
            x_lims = [opts.Extent(3), opts.Extent(4)];
            z_lims = [opts.Extent(1), opts.Extent(2)];
            plt = imagesc(ax, x_lims, z_lims, zeros(size(u)));
        else
            plt = imagesc(ax, zeros(size(u)));
        end
        colormap(ax, opts.ColorMap);
        caxis(ax, clim);
        axis image;
        
        set(ax, 'TickDir', 'out', 'Color', [0.5 0.5 0.5]);
        ax.XAxis.Color = 'k';
        ax.YAxis.Color = 'k';
        
        xlabel(['X (Lateral)' unit_suffix], 'Interpreter', 'latex', 'FontSize', 18);
        ylabel(['Z (Depth)' unit_suffix], 'Interpreter', 'latex', 'FontSize', 18);
        
        if not(isempty(opts.Outline))
            hold(ax, 'on');
            if has_extent && numel(opts.Extent) >= 4
                x_vals_2d = linspace(x_lims(1), x_lims(2), size(u,2));
                z_vals_2d = linspace(z_lims(1), z_lims(2), size(u,1));
                [~, ~] = contour(ax, x_vals_2d, z_vals_2d, opts.Outline, [0.5 0.5], 'Color', 'cyan', 'LineWidth', 2);
            else
                [~, ~] = contour(ax, opts.Outline, [0.5 0.5], 'Color', 'cyan', 'LineWidth', 2);
            end
            hold(ax, 'off');
        end
        
        if not(isempty(opts.Overlay))
            boundaries = bwboundaries(opts.Overlay);
            for i = 1:numel(boundaries)
                b = boundaries{i};
                
                if has_extent && numel(opts.Extent) >= 4
                    x_coords = x_lims(1) + (b(:,2) - 1) * (x_lims(2) - x_lims(1)) / (size(u,2) - 1);
                    z_coords = z_lims(1) + (b(:,1) - 1) * (z_lims(2) - z_lims(1)) / (size(u,1) - 1);
                    patch(ax, x_coords, z_coords, 'r', 'FaceAlpha', 0.3, 'EdgeColor', 'r', 'LineWidth', 2.5);
                else
                    patch(ax, b(:,2), b(:,1), 'r', 'FaceAlpha', 0.3, 'EdgeColor', 'r', 'LineWidth', 2.5);
                end
            end
        end
    end
    
    if not(isempty(ax))
        set(ax, 'FontName', 'Times New Roman', 'FontSize', 14, 'TickLabelInterpreter', 'latex');
        % Sanitize title string to prevent LaTeX crashing on special chars (e.g. °)
        latex_title = strrep(opts.Title, '°', '$^\circ$');
        title(ax, latex_title, 'Color', [0 0 0], 'Interpreter', 'latex', 'FontSize', 24);
        hold on;
    end

    % --- Video Setup ---
    export_video = ~isempty(opts.VideoFilename);
    if export_video
        if endsWith(opts.VideoFilename, '.avi')
            vidObj = VideoWriter(opts.VideoFilename, 'Uncompressed AVI');
        else
            vidObj = VideoWriter(opts.VideoFilename, 'MPEG-4');
            vidObj.Quality = 100;
        end
        vidObj.FrameRate = 60;
        
        open(vidObj);
    end

    % --- Animation Loop ---
    time_steps = 0 : opts.Speed : (maxv + opts.PulseWidth);
    
    for t = time_steps
        % --- Fast Analytical Pulse Calculation ---
        diff_t = u_safe - t;
        
        % Create a mask of only the active wavefront to save computation
        valid_mask = abs(diff_t) <= (opts.PulseWidth / 2);
        
        % Initialize empty image.
        img = complex(zeros(size(u_safe), 'like', u_safe));
        envelope_full = zeros(size(u_safe), 'like', u_safe);
        
        if any(valid_mask, 'all')
            d_valid = diff_t(valid_mask);
            
            % Analytical Hanning envelope: 0.5 * (1 + cos(2*pi * x / width))
            envelope = 0.5 * (1 + cos(2 * pi * d_valid / opts.PulseWidth));
            envelope_full(valid_mask) = envelope;
            
            % Apply envelope and carrier phase.
            img(valid_mask) = envelope .* exp(1i * freq * d_valid / opts.PulseWidth);
        end
        
        img = img .* v_anim;

        % Apply Style.
        if strcmpi(opts.Style, 'real')
            final_img = real(img);
        elseif strcmpi(opts.Style, 'abs')
            final_img = abs(img);
        elseif strcmpi(opts.Style, 'db')
            final_img = 20 * log10(abs(img)) - db_ref;
        end
        
        % Render.
        if is3D
            if strcmpi(opts.RenderMode, 'slice')
                % Dynamically extract slices and cast envelope masks to double
                slice_x = gather(squeeze(final_img(:, sx_idx, :)));
                alpha_x = double(gather(squeeze(envelope_full(:, sx_idx, :))));
                
                % slice() internally transposes planes for certain axes. 
                % We explicitly map our CData/AlphaData shape to match the slice shape.
                if ~isequal(size(slice_x), size(hSlices(1).CData))
                    slice_x = slice_x.';
                    alpha_x = alpha_x.';
                end
                hSlices(1).CData = slice_x;
                set(hSlices(1), 'AlphaData', alpha_x); 
                
                slice_y = gather(squeeze(final_img(sy_idx, :, :)));
                alpha_y = double(gather(squeeze(envelope_full(sy_idx, :, :))));
                
                if ~isequal(size(slice_y), size(hSlices(2).CData))
                    slice_y = slice_y.';
                    alpha_y = alpha_y.';
                end
                hSlices(2).CData = slice_y;
                set(hSlices(2), 'AlphaData', alpha_y); 
                
                if not(isempty(sz_idx))
                    slice_z = gather(squeeze(final_img(:, :, sz_idx)));
                    alpha_z = double(gather(squeeze(envelope_full(:, :, sz_idx))));
                    
                    if ~isequal(size(slice_z), size(hSlices(3).CData))
                        slice_z = slice_z.';
                        alpha_z = alpha_z.';
                    end
                    hSlices(3).CData = slice_z;
                    set(hSlices(3), 'AlphaData', alpha_z); 
                end

            elseif strcmpi(opts.RenderMode, 'isosurface')
                fv = isosurface(X_grid, Y_grid, Z_grid, gather(final_img), iso_val);
                if not(isempty(fv.faces))
                    set(hPatch, 'Faces', fv.faces, 'Vertices', fv.vertices, 'FaceVertexCData', []);
                end
            elseif strcmpi(opts.RenderMode, 'contourslice')
                % Contours must be recreated
                if ~isempty(hContours), delete(hContours); end
                hContours = contourslice(ax, X_grid, Y_grid, Z_grid, gather(final_img), sx, sy, sz, 12);
                set(hContours, 'EdgeColor', 'flat', 'LineWidth', 1.5);
            elseif strcmpi(opts.RenderMode, 'cloud')
                f_img = gather(final_img);
                % Find indices where wave intensity is high enough
                idx = find(abs(f_img) > cloud_thresh);
                % Downsample points to maintain 30+ FPS (stride of 2 means 1/8th the points)
                idx = idx(1:2:end); 
                % Update scatter plot data
                set(hCloud, 'XData', X_grid(idx), 'YData', Y_grid(idx), ...
                            'ZData', Z_grid(idx), 'CData', f_img(idx));
            elseif strcmpi(opts.RenderMode, 'volshow')
                % Normalize data to [0, 1] based on clim_vals for smooth volshow rendering
                norm_img = (final_img - clim(1)) / (clim(2) - clim(1));
                hVol.Data = gather(max(0, min(1, norm_img))); % Clamp and update
            end
            drawnow limitrate;
        else
            plt.CData = gather(final_img);
            set(plt, 'AlphaData', gather(envelope_full));
            drawnow;
        end

        if export_video
            writeVideo(vidObj, getframe(fig));
        end
    end

    if export_video
        close(vidObj);
        disp(['Video saved to ' opts.VideoFilename]);
    end
end